modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -3095,6 +3095,8 @@ def _arrow_type_to_dtype(cls, arrow_type):
         except NotImplementedError:
             if pyarrow.types.is_time(arrow_type):
                 res = np.dtype(datetime.time)
+            elif pyarrow.types.is_dictionary(arrow_type):
+                res = pandas.CategoricalDtype
```

Comment:
`pandas.CategoricalDtype` is just a class and thus is considered as an `np.object`, not an actual type. The test is passing because `df_equals` compares pandas results with [modin results converted to pandas (`mdf._to_pandas()`)](https://github.com/modin-project/modin/blob/a0e5a650b54cc14451dddc8edaf40731ecc212e9/modin/pandas/test/utils.py#L587-L591). 

When a modin dataframe holding pyarrow table is converted to pandas the behavior is literally `return mdf._partitions[0][0].get().to_pandas()` so no metadata comparison of Modin and Pandas dataframes is inquired.

```
Modin.dtypes:
        <class 'numpy.dtype[object_]'>
        dtype('O')

Modin.to_pandas().dtypes:
        <class 'pandas.core.dtypes.dtypes.CategoricalDtype'>
        CategoricalDtype(categories=['first', 'second', 'third'], ordered=False)
```

<details><summary>repro</summary>

```python
from modin.pandas.utils import from_arrow
import pyarrow
import modin.pandas as pd

indices = pyarrow.array([0, 1, 0, 1, 2, 0, None, 2])
dictionary = pyarrow.array(["first", "second", "third"])
dict_array = pyarrow.DictionaryArray.from_arrays(indices, dictionary)
at = pyarrow.table({"col": dict_array})

mdf = from_arrow(at)

md_dtypes = mdf.dtypes[0]
md_to_pandas_dtypes = mdf._to_pandas().dtypes[0]

print(f"Modin.dtypes:\n\t{type(md_dtypes)}\n\t{repr(md_dtypes)}\n")
print(f"Modin.to_pandas().dtypes:\n\t{type(md_to_pandas_dtypes)}\n\t{repr(md_to_pandas_dtypes)}")

```

</details>

So, what this change is doing is just silencing the error of disability to convert `pyarrow.dictionary` to a pandas-compatible type. A proper fix would require materializing the whole column in order to build `pandas.CategoricalDtype` with valid metadata (proper categorical values, is ordered/unordered ...). I believe we should revert the changes in this PR and come up with some other solution to the problem.

Reply:
@dchigarev, good catch!

```
A proper fix would require materializing the whole column in order to build pandas.CategoricalDtype with valid metadata (proper categorical values, is ordered/unordered ...)
```

That is an expensive operation. We could convert arrow table to pandas df in modin_frame.from_arrow() right away in order to get proper dtypes. So there wouldn't be the need in partition_mgr.from_arrow() and modin_frame._arrow_type_to_dtype().